heapprofiler: add periodic query dump mechanism

abarganier opened the pull request:
In an effort to help root-cause OOM errors, there
has been an increase in chatter around the desire
for active query dumps. The hope is that active
query dumps can help SREs and TSEs better correlate
specific queries to OOM issues.

The new queryprofiler uses cgroup memory as a heuristic.
Upon instantiation, it reads the process memory limit
as defined by the process's cgroup. For each periodic
run of queryprofiler, it will compare the current cgroup
mem usage value to that of the previous run. If the rate of
change between these two measurements, when compared
to the cgroup limit, indicates that an OOM is probable
before the next scheduled run, a query dump will be
performed, writing all the currently active queries
from the statusServer's session registry to a CSV
file.

Release note (ops change): operators will potentially
have access to a CSV file containing active queries
from the last run of the queryprofile upon OOM. This
feature is toggled on/off by a new cluster setting,
"diagnostics.active_query_dumps.enabled"

Release justification: Low-risk addition of a new
observability feature providing insight into active
queries when memory pressure is detected.

abarganier committed:
heapprofiler: add periodic query dump mechanism

In an effort to help root-cause OOM errors, there
has been an increase in chatter around the desire
for active query dumps. The hope is that active
query dumps can help SREs and TSEs better correlate
specific queries to OOM issues.

The new queryprofiler uses cgroup memory as a heuristic.
Upon instantiation, it reads the process memory limit
as defined by the process's cgroup. For each periodic
run of queryprofiler, it will compare the current cgroup
mem usage value to that of the previous run. If the rate of
change between these two measurements, when compared
to the cgroup limit, indicates that an OOM is probable
before the next scheduled run, a query dump will be
performed, writing all the currently active queries
from the statusServer's session registry to a CSV
file.

Release note (ops change): operators will potentially
have access to a CSV file containing active queries
from the last run of the queryprofile upon OOM. This
feature is toggled on/off by a new cluster setting,
"diagnostics.active_query_dumps.enabled"

Release justification: Low-risk addition of a new
observability feature providing insight into active
queries when memory pressure is detected.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/66901)
<!-- Reviewable:end -->

joshimhoff reviewed (commented):
Exciting!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66901#-:-MdIMt21AVEZJdHZl48b:b-pxyq17)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @abarganier, @jordanlewis, @knz, and @tbg)

___
*[pkg/server/server.go, line 2413 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66901#-MdIJm6p0BBguPL_b2vI:-MdIJm6p0BBguPL_b2vJ:b-u7i43k) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b1ad3ac514538a2a11f6287c91d3c7b5c2a6f91/pkg/server/server.go#L2413)):*
> ```Go
> }
> 
> type sampleEnvironmentCfg struct {
> ```

Wonderful how small a change this is. Can we backport it eventually? Flagged off if needed?

This ties into a Q I bet you have already thought about: How expensive is dumping queries being executed?
___
*[pkg/server/debug/queries_writer.go, line 12 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66901#-MdIM-_g71nh745LwH_M:-MdIM-_g71nh745LwH_N:b-tvs372) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b1ad3ac514538a2a11f6287c91d3c7b5c2a6f91/pkg/server/debug/queries_writer.go#L12)):*
> ```Go
> )
> 
> var csvHeader = []string{
> ```

Can you document the non-obvious fields? Perhaps all of these are obvious to others but I am confused by:

- Session Alloc. Bytes (which perhaps is accounted for SQL mem associated with a session??)
- SQL Anon.
- Phase
- Progress
___
*[pkg/server/debug/queries_writer.go, line 17 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66901#-MdIMGZr-9wi6rUVGU4i:-MdIMGZr-9wi6rUVGU4j:b-i92rxt) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b1ad3ac514538a2a11f6287c91d3c7b5c2a6f91/pkg/server/debug/queries_writer.go#L17)):*
> ```Go
> 	"TxnID",
> 	"Start Time",
> 	"Session Alloc. Bytes",
> ```

Do we measure allocated bytes associated with a query also?
___
*[pkg/server/heapprofiler/queryprofiler.go, line 43 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66901#-MdIMYRG8enjKWijoUJc:-MdIMYRG8enjKWijoUJd:b9iiuut) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b1ad3ac514538a2a11f6287c91d3c7b5c2a6f91/pkg/server/heapprofiler/queryprofiler.go#L43)):*
> ```Go
> // memory limit.
> //
> // Profiles are also GCed periodically. The latest is always kept, and a couple
> ```

How does this compare to the behavior of the heap profiler?
___
*[pkg/server/heapprofiler/queryprofiler.go, line 44 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66901#-MdIMb_LBGGWGK8W9w2e:-MdIMb_LBGGWGK8W9w2f:bndp6up) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b1ad3ac514538a2a11f6287c91d3c7b5c2a6f91/pkg/server/heapprofiler/queryprofiler.go#L44)):*
> ```Go
> //
> // Profiles are also GCed periodically. The latest is always kept, and a couple
> // of the ones with the largest heap are also kept.
> ```

I understand the logic of this but am not sure about it, as query dumps may be useful for investigating high CPU usage also.
___
*[pkg/server/heapprofiler/queryprofiler.go, line 85 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66901#-MdIKUep9onBr3_Y2x-j:-MdIKUep9onBr3_Y2x-k:bjlaebn) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b1ad3ac514538a2a11f6287c91d3c7b5c2a6f91/pkg/server/heapprofiler/queryprofiler.go#L85)):*
nit: Do we like the term "profile" for this? Makes me think sampling but I'm wrong about that as per this line from wikipedia:

> Some profilers operate by sampling.

https://en.wikipedia.org/wiki/Profiling_(computer_programming)#Profiler_types_based_on_output

Alt suggestion would be `MaybeDumpQueries` or similar. No strong feelings. If I had to argue for not using the term "profile" here, I'd argue avoiding that term may reduce operational confusion about what to look at during a prod incident.
___
*[pkg/server/heapprofiler/queryprofiler.go, line 85 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66901#-MdIKyET-xlxootbucGY:-MdIKyET-xlxootbucGZ:b-bcv4pl) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b1ad3ac514538a2a11f6287c91d3c7b5c2a6f91/pkg/server/heapprofiler/queryprofiler.go#L85)):*
> ```Go
> }
> 
> // MaybeTakeProfile takes a query profile if the rate of change between curRSS
> ```

Do we want synchronization between when a heap profile is taken and when a query profile is taken? Do we already have that synchronization in practice? If yes, do we care to get that synchronization in a way that is more explicit than "the heuristics used to trigger a profile are the same"?
___
*[pkg/server/heapprofiler/queryprofiler.go, line 88 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66901#-MdIKCzs4HhykzAN_rNn:-MdIKCzs4HhykzAN_rNo:b-nahubh) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b1ad3ac514538a2a11f6287c91d3c7b5c2a6f91/pkg/server/heapprofiler/queryprofiler.go#L88)):*
> ```Go
> // and prevRSS indicates that we are close to the system memory limit, implying
> // OOM is probable.
> func (o *QueryProfiler) MaybeTakeProfile(
> ```

These query dumps may be useful if low on CPU also. I wonder if we should periodically dump executing queries, regardless of what is happening on the cluster. I guess could also dump executing queries when CPU usage grows.


<!-- Sent from Reviewable.io -->

joshimhoff reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66901)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @abarganier, @jordanlewis, @knz, and @tbg)

___
*[pkg/server/server.go, line 1690 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66901#-MdISlE23jpcHpiamNnS:-MdISlE23jpcHpiamNnT:bsiyxrx) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b1ad3ac514538a2a11f6287c91d3c7b5c2a6f91/pkg/server/server.go#L1690)):*
> ```Go
> 		heapProfileDirName:   s.cfg.HeapProfileDirName,
> 		runtime:              s.runtime,
> 		sessionRegistry:      s.status.sessionRegistry,
> ```

One other thought (prob not for this PR): Would be nice to be able to manually trigger a query dump, e.g. thru the admin UI.


<!-- Sent from Reviewable.io -->

knz commented:
I love this PR - this is clear and functional (and likely very useful already)
and very well back-portable!

> One other thought (prob not for this PR): Would be nice to be able to manually trigger a query dump, e.g. thru the admin UI.

yes that's not this PR, but I agree we should aim to connect our current "dumper" interface to a HTTP endpoint to enable operators to trigger dumps. Maybe worth filing as a separate project (and ensure this is coherently integrated with our file retrieval API endpoint, to get the results of the dump over the wire too.)

abarganier commented:
Thanks for the feedback, I'll work on addressing comments later today.

In the meantime, I wanted to begin a discussion here about potential PII concerns related to writing these queries to disk. 

In `queries_writer.go` we primarily use the `ActiveQuery` objects (`pkg/server/serverpb/status.proto`) from the session registry to dump info to disk. The proto field we write that contains the actual query is `sql_anon`. I'm curious - how "anonymous" is this field exactly? Is this considered a safely redacted form of the original raw SQL query? Or do PII concerns still remain with this field?

cc @THardy98 from `sql-observability` for their $0.02

Azhng reviewed (commented):
When we surface statement statistics (either on DB Console or crdb_internal views), we anonymize the query by removing all the constant in the query.

For instance:

```
SELECT col1 + 1 FROM table WHERE col2 = 'abc';
```

becomes 

```
SELECT col1 + _ FROM table WHERE col2 = _;
```

I'm not too quite sure if we consider this a safe redaction from a product perspective. cc @kevin-v-ngo 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66901#-:-MdNeLqu5gJGiaH-pmdo:bz1wjkp)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @abarganier, @jordanlewis, @joshimhoff, @knz, and @tbg)

___
*[pkg/server/debug/queries_writer.go, line 17 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66901#-MdIMGZr-9wi6rUVGU4i:-MdNKv_K18ZUu-BSO6Ty:bdpzogi) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b1ad3ac514538a2a11f6287c91d3c7b5c2a6f91/pkg/server/debug/queries_writer.go#L17)):*
<details><summary><i>Previously, joshimhoff (Josh Imhoff) wrote…</i></summary><blockquote>

Do we measure allocated bytes associated with a query also?
</blockquote></details>

We sample bytes associated with a statement in the [`sqlstats`] package. However, the data won't be available until the statement finishes executing. 
___
*[pkg/server/heapprofiler/queryprofiler.go, line 99 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66901#-MdNf_0z45f6QEWXp6Ho:-MdNf_0z45f6QEWXp6Hp:bjzovu3) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b1ad3ac514538a2a11f6287c91d3c7b5c2a6f91/pkg/server/heapprofiler/queryprofiler.go#L99)):*
> ```Go
> 	now := o.now()
> 	diff := curRSS - prevRSS
> 	if curRSS+diff >= (o.maxRSS-varianceBytes) &&
> ```

This is a very interesting strategy to predict a possible OOM scenario. In SQL Observabiliy, we track SQL statement and transaction statistics and flush them to system tables periodically or when we detect there's a memory pressure. 

The heuristics we use to detect memory pressure is through either allocation failure in the memory monitor or when certain limit (defined by the cluster setting) has been exceeded.

This probably does not belong in this PR. I'm wondering if we can expose (and if we have done so in the past) this signal to other part of the system (e.g. our statement/transaction statistics system) so we can also perform certain actions and persist our data before OOM occurs. 


<!-- Sent from Reviewable.io -->

dhartunian reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66901)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @abarganier, @Azhng, @jordanlewis, @joshimhoff, and @knz)

___
*[pkg/server/heapprofiler/queryprofiler.go, line 99 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66901#-MdNf_0z45f6QEWXp6Ho:-MfhYx1K46marXeGn_Rq:byr8mzu) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b1ad3ac514538a2a11f6287c91d3c7b5c2a6f91/pkg/server/heapprofiler/queryprofiler.go#L99)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

This is a very interesting strategy to predict a possible OOM scenario. In SQL Observabiliy, we track SQL statement and transaction statistics and flush them to system tables periodically or when we detect there's a memory pressure. 

The heuristics we use to detect memory pressure is through either allocation failure in the memory monitor or when certain limit (defined by the cluster setting) has been exceeded.

This probably does not belong in this PR. I'm wondering if we can expose (and if we have done so in the past) this signal to other part of the system (e.g. our statement/transaction statistics system) so we can also perform certain actions and persist our data before OOM occurs. 
</blockquote></details>

@azhng can you paste some links to how the memory pressure monitoring is done on the SQL-side? I'm curious to learn more.


<!-- Sent from Reviewable.io -->

abarganier reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66901)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @abarganier, @Azhng, @jordanlewis, @joshimhoff, and @knz)

___
*[pkg/server/heapprofiler/queryprofiler.go, line 43 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66901#-MdIMYRG8enjKWijoUJc:-Mfi5eco20FsR4AuAsX8:b4pz27s) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b1ad3ac514538a2a11f6287c91d3c7b5c2a6f91/pkg/server/heapprofiler/queryprofiler.go#L43)):*
<details><summary><i>Previously, joshimhoff (Josh Imhoff) wrote…</i></summary><blockquote>

How does this compare to the behavior of the heap profiler?
</blockquote></details>

Same behavior :) 
___
*[pkg/server/heapprofiler/queryprofiler.go, line 99 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66901#-MdNf_0z45f6QEWXp6Ho:-Mfi5wV67QeJQ5CqUufO:bh9l558) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b1ad3ac514538a2a11f6287c91d3c7b5c2a6f91/pkg/server/heapprofiler/queryprofiler.go#L99)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

@azhng can you paste some links to how the memory pressure monitoring is done on the SQL-side? I'm curious to learn more.
</blockquote></details>

With some experimentation and testing, while the idea behind this strategy is valid, it's being done incorrectly here. System RSS is generally way higher than what a container considers as "used memory". For example, if CRDB is running in a Docker container, `docker stats` will show memory usage of the container consistently 40-60% lower than this system RSS measurement. We should find a more accurate measurement that allows us to be more conservative when determining whether or not to dump queries.

I think a better measurement would be to use the memory cgroup to read current memory usage, and then compare that to the cgroup memory limit. Most [containers consider something along the lines of](https://docs.docker.com/engine/reference/commandline/stats/#extended-description) `mem_usage - cache_usage` when considering current memory usage. So, in a separate PR I'm going to expand our cgroups library to allow us to read these values, which we can then fetch on a scheduled interval. I'll link that here once it's up.


<!-- Sent from Reviewable.io -->

Azhng reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66901)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @abarganier, @dhartunian, @jordanlewis, @joshimhoff, and @knz)

___
*[pkg/server/heapprofiler/queryprofiler.go, line 99 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66901#-MdNf_0z45f6QEWXp6Ho:-MgCRpXM8-fE2ZIDH0wG:ba325t6) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b1ad3ac514538a2a11f6287c91d3c7b5c2a6f91/pkg/server/heapprofiler/queryprofiler.go#L99)):*
<details><summary><i>Previously, abarganier (Alex Barganier) wrote…</i></summary><blockquote>

With some experimentation and testing, while the idea behind this strategy is valid, it's being done incorrectly here. System RSS is generally way higher than what a container considers as "used memory". For example, if CRDB is running in a Docker container, `docker stats` will show memory usage of the container consistently 40-60% lower than this system RSS measurement. We should find a more accurate measurement that allows us to be more conservative when determining whether or not to dump queries.

I think a better measurement would be to use the memory cgroup to read current memory usage, and then compare that to the cgroup memory limit. Most [containers consider something along the lines of](https://docs.docker.com/engine/reference/commandline/stats/#extended-description) `mem_usage - cache_usage` when considering current memory usage. So, in a separate PR I'm going to expand our cgroups library to allow us to read these values, which we can then fetch on a scheduled interval. I'll link that here once it's up.
</blockquote></details>

@dhartunian we initially introduced memory accounting in #65172. Then in #67866 we starts to [emit](https://github.com/cockroachdb/cockroach/blob/55b15b04ffe3179fa7b5d3131b11ffc877cdd629/pkg/sql/sqlstats/persistedsqlstats/writer.go#L72) memory pressure signal via a Go channel. 


<!-- Sent from Reviewable.io -->

abarganier reviewed (commented):
Thanks for the feedback everyone, this is ready for review now that I've implemented the cgroup memory stats into the heuristic. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66901#-:-MhKIv521V12q6ZlD3Y_:b-16aadk)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @abarganier, @dhartunian, @jordanlewis, @joshimhoff, and @knz)

___
*[pkg/server/server.go, line 1690 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66901#-MdISlE23jpcHpiamNnS:-MhEO_yU-CQAwqYZRRSo:b-yqi578) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b1ad3ac514538a2a11f6287c91d3c7b5c2a6f91/pkg/server/server.go#L1690)):*
<details><summary><i>Previously, joshimhoff (Josh Imhoff) wrote…</i></summary><blockquote>

One other thought (prob not for this PR): Would be nice to be able to manually trigger a query dump, e.g. thru the admin UI.
</blockquote></details>

Yes agreed, I'd be happy to add a hook for this. Is it different enough from what the "Statements" page provides though to be worth it? I suppose it would provide a more durable record... 
___
*[pkg/server/server.go, line 2413 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66901#-MdIJm6p0BBguPL_b2vI:-MhEnfr70pStG92A9dZu:bmqhas6) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b1ad3ac514538a2a11f6287c91d3c7b5c2a6f91/pkg/server/server.go#L2413)):*
<details><summary><i>Previously, joshimhoff (Josh Imhoff) wrote…</i></summary><blockquote>

Wonderful how small a change this is. Can we backport it eventually? Flagged off if needed?

This ties into a Q I bet you have already thought about: How expensive is dumping queries being executed?
</blockquote></details>

I believe this change should be easily backportable, should we put it behind a flag regardless though?

I think the main cost with dumping queries is the serialization of the SessionRegistry. If we decide we want to dump, we have to lock the session registry and serialize each active session into a `[]serverpb.Session`. The overhead for this I assume to be acceptable, since the same action is performed every time someone accesses the "Sessions" page in DB Console, which then does periodic polls every ~10 seconds which again triggers the action. 
___
*[pkg/server/debug/queries_writer.go, line 12 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66901#-MdIM-_g71nh745LwH_M:-MhEohgR0UQgSPY0TUO0:b-wufyiz) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b1ad3ac514538a2a11f6287c91d3c7b5c2a6f91/pkg/server/debug/queries_writer.go#L12)):*
<details><summary><i>Previously, joshimhoff (Josh Imhoff) wrote…</i></summary><blockquote>

Can you document the non-obvious fields? Perhaps all of these are obvious to others but I am confused by:

- Session Alloc. Bytes (which perhaps is accounted for SQL mem associated with a session??)
- SQL Anon.
- Phase
- Progress
</blockquote></details>

Good call, I went ahead and documented all of the fields.
___
*[pkg/server/heapprofiler/queryprofiler.go, line 44 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66901#-MdIMb_LBGGWGK8W9w2e:-MhEQdSE1egzEqs1jAjj:b-rifq2i) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b1ad3ac514538a2a11f6287c91d3c7b5c2a6f91/pkg/server/heapprofiler/queryprofiler.go#L44)):*
<details><summary><i>Previously, joshimhoff (Josh Imhoff) wrote…</i></summary><blockquote>

I understand the logic of this but am not sure about it, as query dumps may be useful for investigating high CPU usage also.
</blockquote></details>

It follows the same standard logic used by the rest of the heapprofiler system, so I'm hesitant in this first pass to make any tweaks to the strategy as it feels like it grows the scope of the PR quite a bit. I'd be happy to work to improve this over time though. 

Just curious, what kind of strategy do you think would be more appropriate?  Sounds like we should also consider the profiles which were taken with the highest CPU usage?
___
*[pkg/server/heapprofiler/queryprofiler.go, line 85 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66901#-MdIKUep9onBr3_Y2x-j:-MhEdMN1-bkQFZENZNkf:bw5ialm) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b1ad3ac514538a2a11f6287c91d3c7b5c2a6f91/pkg/server/heapprofiler/queryprofiler.go#L85)):*
<details><summary><i>Previously, joshimhoff (Josh Imhoff) wrote…</i></summary><blockquote>

nit: Do we like the term "profile" for this? Makes me think sampling but I'm wrong about that as per this line from wikipedia:

> Some profilers operate by sampling.

https://en.wikipedia.org/wiki/Profiling_(computer_programming)#Profiler_types_based_on_output

Alt suggestion would be `MaybeDumpQueries` or similar. No strong feelings. If I had to argue for not using the term "profile" here, I'd argue avoiding that term may reduce operational confusion about what to look at during a prod incident.
</blockquote></details>

Yeah I agree, switched up the verbage here. 
___
*[pkg/server/heapprofiler/queryprofiler.go, line 85 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66901#-MdIKyET-xlxootbucGY:-MhEdg7yB8x65kea5fr3:b-9aqn7b) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b1ad3ac514538a2a11f6287c91d3c7b5c2a6f91/pkg/server/heapprofiler/queryprofiler.go#L85)):*
<details><summary><i>Previously, joshimhoff (Josh Imhoff) wrote…</i></summary><blockquote>

Do we want synchronization between when a heap profile is taken and when a query profile is taken? Do we already have that synchronization in practice? If yes, do we care to get that synchronization in a way that is more explicit than "the heuristics used to trigger a profile are the same"?
</blockquote></details>

I agree this would be nice to have, but what would we synchronize? Wouldn't synchronization here from one profile (e.g. the heapprofiler) to the next essentially require us to stop-the-world to prevent DB node state changes? Let me know if I'm misunderstanding the question :sweat_smile: 
___
*[pkg/server/heapprofiler/queryprofiler.go, line 88 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66901#-MdIKCzs4HhykzAN_rNn:-MhEgQKVCo_eQ7x6gBdM:ba9v1u3) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b1ad3ac514538a2a11f6287c91d3c7b5c2a6f91/pkg/server/heapprofiler/queryprofiler.go#L88)):*
<details><summary><i>Previously, joshimhoff (Josh Imhoff) wrote…</i></summary><blockquote>

These query dumps may be useful if low on CPU also. I wonder if we should periodically dump executing queries, regardless of what is happening on the cluster. I guess could also dump executing queries when CPU usage grows.
</blockquote></details>

I like the idea of expanding upon the heuristic used to trigger a dump - we could certainly add some CPU usage stats to the equation in the future.

With respect to dumping regardless, sounds like something that we could put behind a cluster setting that could be explicitly enabled?


<!-- Sent from Reviewable.io -->

jordanlewis commented:
@Azhng please weigh in on the above questions when you have a chance, thanks!

knz reviewed (commented):
I'm generally OK with the approach. Deferring to Josh about the heuristic.

This PR also needs a unit test.

> Reviewed 3 of 3 files at r3, 2 of 2 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66901#-:-MhY7men3k0RYLOt3HsZ:b-6skow6)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @abarganier, @dhartunian, @jordanlewis, and @joshimhoff)

___
*[pkg/server/server.go, line 1690 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66901#-MdISlE23jpcHpiamNnS:-MhY6LZG-lbyww5YIwmP:bvvf6j6) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b1ad3ac514538a2a11f6287c91d3c7b5c2a6f91/pkg/server/server.go#L1690)):*
<details><summary><i>Previously, abarganier (Alex Barganier) wrote…</i></summary><blockquote>

Yes agreed, I'd be happy to add a hook for this. Is it different enough from what the "Statements" page provides though to be worth it? I suppose it would provide a more durable record... 
</blockquote></details>

maybe file a separate issue so we can discuss the mechanism of action and the effect.
___
*[pkg/server/server.go, line 2553 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/66901#-MhY6V2-2FSdDnopIJwp:-MhY6V2-2FSdDnopIJwq:b-x281ro) ([raw file](https://github.com/cockroachdb/cockroach/blob/af43b0ba10ed2814684397c167833792086d0232/pkg/server/server.go#L2553)):*
> ```Go
> 			queryProfiler, err = heapprofiler.NewActiveQueryProfiler(ctx, cfg.heapProfileDirName, cfg.st)
> 			if err != nil {
> 				log.Warningf(ctx, "Failed to start query profiler worker: %s", err.Error())
> ```

- use `%v` for errors.
- pass `err` and don't call `.Error` to ensure the log processor has access to the internal structure.
- no capital letter at start of log messages.
___
*[pkg/server/server.go, line 2614 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/66901#-MhY6l058jRbTmlVZLYs:-MhY6l06AaB6_jEHFH5V:b-xabbpv) ([raw file](https://github.com/cockroachdb/cockroach/blob/af43b0ba10ed2814684397c167833792086d0232/pkg/server/server.go#L2614)):*
> ```Go
> 					nonGoAllocProfiler.MaybeTakeProfile(ctx, cfg.runtime.CgoTotalBytes.Value())
> 					statsProfiler.MaybeTakeProfile(ctx, cfg.runtime.RSSBytes.Value(), curStats, cgoStats)
> 					if queryProfiler != nil {
> ```

are you sure this needs to be inside the heapProfiler conditional? how is this different from goroutinedumper above?
___
*[pkg/server/heapprofiler/queryprofiler.go, line 43 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66901#-MdIMYRG8enjKWijoUJc:-MhY78B7ABpZcw9o4bdm:bjwa7hh) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b1ad3ac514538a2a11f6287c91d3c7b5c2a6f91/pkg/server/heapprofiler/queryprofiler.go#L43)):*
<details><summary><i>Previously, abarganier (Alex Barganier) wrote…</i></summary><blockquote>

Same behavior :) 
</blockquote></details>

:tada: 


<!-- Sent from Reviewable.io -->

Azhng reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66901)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @abarganier, @dhartunian, @jordanlewis, and @joshimhoff)

___
*[pkg/server/server.go, line 1690 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66901#-MdISlE23jpcHpiamNnS:-MhpE0S16Lu2gAPFfh_R:b-k6ovxj) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b1ad3ac514538a2a11f6287c91d3c7b5c2a6f91/pkg/server/server.go#L1690)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

maybe file a separate issue so we can discuss the mechanism of action and the effect.
</blockquote></details>

"Statements" Page is currently powered by the `sqlstats` subsystem. Since we only record the statements statistics after the execution of that statement is complete, any active query will not show up in the "Statement" Page. 
___
*[pkg/server/heapprofiler/queryprofiler.go, line 88 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66901#-MdIKCzs4HhykzAN_rNn:-MhpF0VG0GYMst9rBjL3:b-55sh6t) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b1ad3ac514538a2a11f6287c91d3c7b5c2a6f91/pkg/server/heapprofiler/queryprofiler.go#L88)):*
<details><summary><i>Previously, abarganier (Alex Barganier) wrote…</i></summary><blockquote>

I like the idea of expanding upon the heuristic used to trigger a dump - we could certainly add some CPU usage stats to the equation in the future.

With respect to dumping regardless, sounds like something that we could put behind a cluster setting that could be explicitly enabled?
</blockquote></details>

SQL Observability is introducing [persisted sql stats](#64743) in 21.2, where we can lookup historical query statistics. Perhaps instead of dumping the query, we can work on correlating CPU metrics (e.g. CPU usage spikes) with the the queries that were executing during the same time window? This would enable custom to self-troubleshoot the problem directly? 


<!-- Sent from Reviewable.io -->

abarganier reviewed (commented):
Thanks for the feedback everyone. I've addressed comments and added some unit tests around the heuristic as well as the constructor function. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66901#-:-MhteGhI9_nFwlwfA3wW:bt06qfp)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @abarganier, @dhartunian, @jordanlewis, @joshimhoff, and @knz)

___
*[pkg/server/server.go, line 1690 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66901#-MdISlE23jpcHpiamNnS:-MhswgGA-lwAWp5-5I9K:bt8ysek) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b1ad3ac514538a2a11f6287c91d3c7b5c2a6f91/pkg/server/server.go#L1690)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

"Statements" Page is currently powered by the `sqlstats` subsystem. Since we only record the statements statistics after the execution of that statement is complete, any active query will not show up in the "Statement" Page. 
</blockquote></details>

Good to know, thanks for clarifying Archer. I've filed https://github.com/cockroachdb/cockroach/issues/69307 where we can discuss this proposal in more detail. 
___
*[pkg/server/server.go, line 2553 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/66901#-MhY6V2-2FSdDnopIJwp:-Mhsydob3C7jAs7jLDop:b-x3zn47) ([raw file](https://github.com/cockroachdb/cockroach/blob/af43b0ba10ed2814684397c167833792086d0232/pkg/server/server.go#L2553)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

- use `%v` for errors.
- pass `err` and don't call `.Error` to ensure the log processor has access to the internal structure.
- no capital letter at start of log messages.

</blockquote></details>

:thumbsup: Done.
___
*[pkg/server/server.go, line 2614 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/66901#-MhY6l058jRbTmlVZLYs:-Mhsz-MD-W1QGTAvbKse:b-psqo6d) ([raw file](https://github.com/cockroachdb/cockroach/blob/af43b0ba10ed2814684397c167833792086d0232/pkg/server/server.go#L2614)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

are you sure this needs to be inside the heapProfiler conditional? how is this different from goroutinedumper above?
</blockquote></details>

Ah yes good point, my mistake. I moved this conditional to be at the same level as the goroutinedumper/heapprofiler conditionals. 
___
*[pkg/server/heapprofiler/queryprofiler.go, line 88 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66901#-MdIKCzs4HhykzAN_rNn:-Mht-DzC1_EUUecMrZj0:b-a07fhn) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b1ad3ac514538a2a11f6287c91d3c7b5c2a6f91/pkg/server/heapprofiler/queryprofiler.go#L88)):*
I like this idea - I think given all the metrics we have flying around, the "always dump" idea feels a bit verbose and could lead to excess "clutter", making it difficult to know which files are noteworthy. 

> instead of dumping the query

Just clarifying - instead of dumping are you proposing keeping a record of "Query X occurred during CPU spike Y" somewhere else, and surfacing that information in a UI? I'd be down to file a separate issue to develop this idea more! Let me know your thoughts.

Regarding the query dump heuristic itself, I'd like to keep it memory-specific within the scope of this PR but I am into the idea of expanding upon the heuristic. I've filed https://github.com/cockroachdb/cockroach/issues/69309 to discuss further, and also noted that it would be nice to expose this heuristic/trigger to other parts of the system through some new abstraction. 


<!-- Sent from Reviewable.io -->

joshimhoff reviewed (commented):
LGTM! Very exciting. I made one "feature suggestion" that may be too late to execute on, given when 21.2 branch will be cut.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66901#-:-MhtsGDT1OAPlMUYOsse:bdves3a)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @abarganier, @dhartunian, @jordanlewis, @joshimhoff, and @knz)

___
*[pkg/server/server.go, line 2413 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66901#-MdIJm6p0BBguPL_b2vI:-MhtqyDd47J7l16bBH74:b-5lf9j1) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b1ad3ac514538a2a11f6287c91d3c7b5c2a6f91/pkg/server/server.go#L2413)):*
<details><summary><i>Previously, abarganier (Alex Barganier) wrote…</i></summary><blockquote>

I believe this change should be easily backportable, should we put it behind a flag regardless though?

I think the main cost with dumping queries is the serialization of the SessionRegistry. If we decide we want to dump, we have to lock the session registry and serialize each active session into a `[]serverpb.Session`. The overhead for this I assume to be acceptable, since the same action is performed every time someone accesses the "Sessions" page in DB Console, which then does periodic polls every ~10 seconds which again triggers the action. 
</blockquote></details>

If was me, I might provide a cluster setting kill switch or similar, but depends on how risky you think dumping queries is; I think you are in the best position to judge risk of a problem with this feature blowing up a cluster or similar.
___
*[pkg/server/heapprofiler/queryprofiler.go, line 44 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66901#-MdIMb_LBGGWGK8W9w2e:-Mhtr_aEBWUYWEnlzLu1:bytjc7c) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b1ad3ac514538a2a11f6287c91d3c7b5c2a6f91/pkg/server/heapprofiler/queryprofiler.go#L44)):*
<details><summary><i>Previously, abarganier (Alex Barganier) wrote…</i></summary><blockquote>

It follows the same standard logic used by the rest of the heapprofiler system, so I'm hesitant in this first pass to make any tweaks to the strategy as it feels like it grows the scope of the PR quite a bit. I'd be happy to work to improve this over time though. 

Just curious, what kind of strategy do you think would be more appropriate?  Sounds like we should also consider the profiles which were taken with the highest CPU usage?
</blockquote></details>

Longer term: I think a query dump may be useful to trigger in ~four situations:

1. When mem grows or is high for a sustained period of time.
2. When CPU grows or is high for a sustained period of time.
3. Every N seconds (with N being a cluster setting).
4. When user triggers one thru admin UI.

The only one I could see an argument to add in this PR is 3 as it is easy and useful. But happy to be told "no not now; it's too late; 21.2 branch about to be cut". I just think having a query dump every N seconds will be useful even if N~10m. Wdyt about that? Again, NO pressure on adding this functionality so close to 21.2 branch cut time.
___
*[pkg/server/heapprofiler/queryprofiler.go, line 85 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66901#-MdIKyET-xlxootbucGY:-Mhts2-eC8x_2wl9qgIx:bcenrze) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b1ad3ac514538a2a11f6287c91d3c7b5c2a6f91/pkg/server/heapprofiler/queryprofiler.go#L85)):*
<details><summary><i>Previously, abarganier (Alex Barganier) wrote…</i></summary><blockquote>

I agree this would be nice to have, but what would we synchronize? Wouldn't synchronization here from one profile (e.g. the heapprofiler) to the next essentially require us to stop-the-world to prevent DB node state changes? Let me know if I'm misunderstanding the question :sweat_smile: 
</blockquote></details>

Whoops. I must have been tired lol and forgot about how computers work.


<!-- Sent from Reviewable.io -->

abarganier reviewed (commented):
Thanks for the suggestions Josh!  I've added a cluster setting to toggle the feature on/off and added your heuristic suggestions to #69309 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66901#-:-Mhy9-vtEEaUUdVMmbkp:b-i7cm6f)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @abarganier, @dhartunian, @jordanlewis, @joshimhoff, and @knz)

___
*[pkg/server/server.go, line 2413 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66901#-MdIJm6p0BBguPL_b2vI:-Mhy7_322DuXOoKLawoX:b7lkfia) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b1ad3ac514538a2a11f6287c91d3c7b5c2a6f91/pkg/server/server.go#L2413)):*
<details><summary><i>Previously, joshimhoff (Josh Imhoff) wrote…</i></summary><blockquote>

If was me, I might provide a cluster setting kill switch or similar, but depends on how risky you think dumping queries is; I think you are in the best position to judge risk of a problem with this feature blowing up a cluster or similar.
</blockquote></details>

Yeah I agree that a cluster setting would be a nice safety feature here. I went ahead and added `diagnostics.active_query_dumps.enabled` to toggle the feature.
___
*[pkg/server/heapprofiler/queryprofiler.go, line 44 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66901#-MdIMb_LBGGWGK8W9w2e:-Mhy7qza3DEQMvAdgBbO:bpag1uk) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b1ad3ac514538a2a11f6287c91d3c7b5c2a6f91/pkg/server/heapprofiler/queryprofiler.go#L44)):*
<details><summary><i>Previously, joshimhoff (Josh Imhoff) wrote…</i></summary><blockquote>

Longer term: I think a query dump may be useful to trigger in ~four situations:

1. When mem grows or is high for a sustained period of time.
2. When CPU grows or is high for a sustained period of time.
3. Every N seconds (with N being a cluster setting).
4. When user triggers one thru admin UI.

The only one I could see an argument to add in this PR is 3 as it is easy and useful. But happy to be told "no not now; it's too late; 21.2 branch about to be cut". I just think having a query dump every N seconds will be useful even if N~10m. Wdyt about that? Again, NO pressure on adding this functionality so close to 21.2 branch cut time.
</blockquote></details>

I like these requests - I've created #69309 to track/discuss how to expand the heuristic. Aside from putting the feature behind a cluster setting, I'm going to avoid expanding things any further for now due to the branch cut time being close. I look forward to continuing work on this, though. I appreciate the suggestions and have noted them in the issue :) 
___
*[pkg/server/heapprofiler/queryprofiler.go, line 85 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66901#-MdIKyET-xlxootbucGY:-Mhy8vlU3GgiJITwEYS9:bcy1567) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b1ad3ac514538a2a11f6287c91d3c7b5c2a6f91/pkg/server/heapprofiler/queryprofiler.go#L85)):*
<details><summary><i>Previously, joshimhoff (Josh Imhoff) wrote…</i></summary><blockquote>

Whoops. I must have been tired lol and forgot about how computers work.
</blockquote></details>

Hah all good!


<!-- Sent from Reviewable.io -->

dhartunian reviewed (commented):
> Release note (ops change): operators will potentially
> have access to a CSV file containing active queries
> from the last run of the queryprofile upon OOM.

nit: Can you amend the release note to include the new cluster setting?

Also, you'll need to add a release justification. Category 4 sounds right to me.

> Reviewed 1 of 5 files at r5, 5 of 6 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66901#-:-Mi3fZP_6dDyL7v-gqt5:bbtpjkd)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @abarganier, @dhartunian, @jordanlewis, @joshimhoff, and @knz)

___
*[docs/generated/settings/settings.html, line 13 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/66901#-Mi3dLuZDgayrDZHUxMz:-Mi3dLuZDgayrDZHUxN-:btg2ej6) ([raw file](https://github.com/cockroachdb/cockroach/blob/60b92fb1a57094bb470535960c511054b73195d3/docs/generated/settings/settings.html#L13)):*
> ```HTML
> <tr><td><code>cluster.organization</code></td><td>string</td><td><code></code></td><td>organization name</td></tr>
> <tr><td><code>cluster.preserve_downgrade_option</code></td><td>string</td><td><code></code></td><td>disable (automatic or manual) cluster version upgrade from the specified version until reset</td></tr>
> <tr><td><code>diagnostics.active_query_dumps.enabled</code></td><td>boolean</td><td><code>true</code></td><td>enable dumping of anonymized active queries to disk when node is under pressure</td></tr>
> ```

nit: Should we mark this as experimental either with `experimental` in the setting name or in the description?
nit: Also can you amend to clarify "memory pressure" to be specific here?
___
*[pkg/server/heapprofiler/cluster_settings.go, line 17 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/66901#-Mi3fq886CMRb_Yq-y_V:-Mi3fq886CMRb_Yq-y_W:bkadt8r) ([raw file](https://github.com/cockroachdb/cockroach/blob/60b92fb1a57094bb470535960c511054b73195d3/pkg/server/heapprofiler/cluster_settings.go#L17)):*
> ```Go
> // ActiveQueryDumpsEnabled wraps "diagnostics.active_query_dumps.enabled"
> //
> // diagnostics.active_query_dumps.enabled" enables the periodic writing of
> ```

nit: dangling quote mark


<!-- Sent from Reviewable.io -->

abarganier reviewed (commented):
Thanks for the feedback - addressed your comments and updates the release note + added a release justification. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66901#-:-Mi7DP9G0Rj8DRYexZTF:bu0nw5u)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @abarganier, @dhartunian, @jordanlewis, @joshimhoff, and @knz)

___
*[docs/generated/settings/settings.html, line 13 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/66901#-Mi3dLuZDgayrDZHUxMz:-Mi7DKSw8Zuq_3dxzX5C:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/60b92fb1a57094bb470535960c511054b73195d3/docs/generated/settings/settings.html#L13)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

nit: Should we mark this as experimental either with `experimental` in the setting name or in the description?
nit: Also can you amend to clarify "memory pressure" to be specific here?
</blockquote></details>

Done.
___
*[pkg/server/heapprofiler/cluster_settings.go, line 17 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/66901#-Mi3fq886CMRb_Yq-y_V:-Mi7DMyl9Z8iYPR3GR1S:b-x76dwb) ([raw file](https://github.com/cockroachdb/cockroach/blob/60b92fb1a57094bb470535960c511054b73195d3/pkg/server/heapprofiler/cluster_settings.go#L17)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

nit: dangling quote mark
</blockquote></details>

Oops - fixed.


<!-- Sent from Reviewable.io -->

joshimhoff reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

Leaving approval for someone on DB eng.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66901#-:-MiMnS597zHGlgQaSqS1:bpp4hyw)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @abarganier, @dhartunian, @jordanlewis, @joshimhoff, and @knz)



<!-- Sent from Reviewable.io -->

abarganier commented:
Thanks Josh! @knz would you be able to take a look?

rimadeodhar reviewed (approved):
> Reviewed 2 of 3 files at r1, 3 of 3 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66901)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @abarganier, @dhartunian, @jordanlewis, @joshimhoff, and @knz)



<!-- Sent from Reviewable.io -->

abarganier commented:
bors r=rimadeodhar
